Solutions/Corvus.SpecFlow.Extensions/Corvus/SpecFlow/Extensions/ContainerBeforeScenarioOrder.cs

```diff
@@ -0,0 +1,47 @@
+﻿// <copyright file="ContainerBeforeScenarioOrder.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.SpecFlow.Extensions
+{
+    using System;
+    using TechTalk.SpecFlow;
+
+    /// <summary>
+    /// Defines constants to use for <see cref="HookAttribute.Order"/> property on a
+    /// <see cref="BeforeScenarioAttribute"/> to ensure that initialization occurs in the correct order when using
+    /// <see cref="ContainerBindings"/> at scenario scope. Tests must specify the <c>@perScenarioContainer</c> tag to opt
+    /// into this functionality.
+    /// </summary>
+    public static class ContainerBeforeScenarioOrder
+    {
+        /// <summary>
+        /// The <c>Order</c> at which the <c>ServiceCollection</c> for the container is created.
+        /// </summary>
+        public const int CreateServiceCollection = 0;
+
+        /// <summary>
+        /// The minimum <c>Order</c> to specify on <c>BeforeScenario</c> bindings that use
+        /// <see cref="ContainerBindings.ConfigureServices(FeatureContext, Action{Microsoft.Extensions.DependencyInjection.IServiceCollection})"/>.
+        /// </summary>
+        /// <remarks>
+        /// If you need to run multiple bindings that populate the <c>ServiceCollection</c>, and
+        /// if you need to control their order, you can use any number greater than or equal to this,
+        /// and less then <see cref="BuildServiceProvider"/>.
+        /// </remarks>
+        public const int PopulateServiceCollection = CreateServiceCollection + 1;
+
+        /// <summary>
+        /// The <c>Order</c> at which the <c>IServiceProvider</c> gets built from the
+        /// <c>ServiceCollection</c> that was available from when the order reached
+        /// <see cref="PopulateServiceCollection"/>.
+        /// </summary>
+        public const int BuildServiceProvider = 10000;
+
+        /// <summary>
+        /// The <c>Order</c> at which the <c>IServiceProvider</c> built during the
+        /// <see cref="PopulateServiceCollection"/> becomes available.
+        /// </summary>
+        public const int ServiceProviderAvailable = BuildServiceProvider + 1;
+    }
+}
```

Comment:
This file is pretty much identical to `ContainerBeforeFeatureOrder`. Is there really any benefit in separating them out?

Reply:
The main purpose of these classes is so that you can put meaningful names in your `Order` settings. This means the type name is actually a significant part of the "functionality", insofar as this type has any functionality. (The idea is that when you set your `Order` to one of these values, the code describes what your intent was.) So I was OK with this being different only in name because its name is most of its reason for existing.

I also quite liked the fact that the existence of two types here emphasises the fact that these orders exist in two distinct processes. If we had a single type used for both `BeforeScenario` and `BeforeFeature` it gives a false impression that these are basically the same thing. They're not. Every `BeforeFeature` runs before any `BeforeScenario`, so I don't particularly like the idea that you could have two attibutes both specifying an `Order` from this class that don't actually run in the order specified by this class (because you've used one at the feature level and one at the scenario level).

If we were to do what you suggest, I think we'd have to rename the existing type. (At least, I think it'd be a bad idea to use the existing type for the new scenario. That would force `BeforeScenario` attributes to use a type with `BeforeFeature` in it.) That would be a breaking change, which I'm not too upset about given that almost nothing uses this yet. But what would we call it? I can't just take out the `Feature` because the name would then make no sense. I guess it could be `ContainerBindingOrder`? Although that then loses site of the fact that these things are for use specifially in `BeforeScenario` and `BeforeFeature` bindings. (And that's really another manifestation of the fact that I like that with 2 classes today, we currently have a correspondence between the kind of binding attribute you use, and the class you use in your `Order` properties.)

This would essentially be making the assumption that we'll never want significantly different sequencing for scenarios and features here. But I'm not too worried about that, because as far as I can see, it's probably true.

So...on balance I still prefer it how it is. Have my arguments swayed you at all?

Reply:
Yep!